src/mongocxx/instance.cpp

```diff
@@ -94,31 +96,25 @@ instance::instance() : instance(nullptr) {
 }
 
 instance::instance(std::unique_ptr<logger> logger) {
-    std::lock_guard<std::recursive_mutex> lock(instance_mutex);
-    if (current_instance) {
+    instance* expected = nullptr;
+    if (!current_instance.compare_exchange_strong(expected, this))
         throw logic_error{error_code::k_instance_already_exists};
-    }
     _impl = stdx::make_unique<impl>(std::move(logger));
-    current_instance = this;
 }
 
 instance::instance(instance &&) noexcept = default;
 instance &instance::operator=(instance &&) noexcept = default;
 
 instance::~instance() {
-    std::lock_guard<std::recursive_mutex> lock(instance_mutex);
-    if (current_instance != this) std::abort();
     _impl.reset();
-    current_instance = nullptr;
+    current_instance.store(nullptr);
```

Comment:
And, actually... it can't be. There is no proper sequencing of the reset on the _impl and nulling out the current_instance pointer. I think this may be the end for `instance::current`